---
title: "Whole genome comparison of maize assemblies"
output: pdf_document
classoption: landscape
# author:
# - name: Peng Zhou
#  affiliation: University of Minnesota
# - name: Nathan Springer
#  affiliation: University of Minnesota
# abstract: "This document provides an introduction to R Markdown, argues for its..."
# keywords: "pandoc, r markdown, knitr"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
# spacing: double
# bibliography: ~/Dropbox/master.bib
# biblio-style: apsr
header-includes:
- \usepackage{caption}
- \usepackage{booktabs}
- \usepackage{longtable}
- \usepackage{array}
- \usepackage{multirow}
- \usepackage[table]{xcolor}
- \usepackage{wrapfig}
- \usepackage{float}
- \usepackage{colortbl}
- \usepackage{pdflscape}
- \usepackage{tabu}
- \usepackage{threeparttable}
- \usepackage{threeparttablex}
- \usepackage[normalem]{ulem}
- \usepackage{makecell}
- \captionsetup[figure]{labelformat=empty}
- \captionsetup[table]{labelformat=empty}
- \AtBeginDocument{\let\maketitle\relax}
---
\pagenumbering{gobble}

- Pipeline components:
  - [minimap2](https://github.com/lh3/minimap2) was used to align one assembly to another;
  - [blat chain/net tools](http://genomewiki.ucsc.edu/index.php/Whole_genome_alignment_howto) were used to process alignment resuls and build synteny chains/nets;
  - [bcftools](https://samtools.github.io/bcftools/) and [GATK4](https://software.broadinstitute.org/gatk/gatk4) were used to call variants;
  - [snpEff](http://snpeff.sourceforge.net) was used to evaluate variant effects on sytenic genes

- whole genome synteny plots
  - [Mo17 vs B73](/data/07_dotplot/Mo17_B73.pdf)
  - [W22 vs B73](/data/07_dotplot/W22_B73.pdf)
  - [PH207 vs B73](/data/07_dotplot/PH207_B73.pdf)
  - [PHB47 vs B73](/data/07_dotplot/PHB47_B73.pdf)

- Table 1.  Whole genome alignment statistics.
```{r echo = FALSE}
require(tidyverse)
require(knitr)
require(kableExtra)
options(knitr.kable.NA = '')
dirw = '~/projects/wgc/data' 
fi = file.path(dirw, '05_stats', "all.rda")
x = load(fi)
kable(tt, format = 'latex', booktabs = T, 
    align = c('l',rep('r',ncol(tt)-1)),
    col.names = tt_h0) %>%
    #caption = "Table 1.  Alignment statistics.") %>%
  column_spec(1, bold=T) %>%#, width_min='7cm') %>%
  kable_styling(latex_options = c('striped','scale_down')) %>%
  add_header_above(c('', tt_h1)) %>%
  add_header_above(c('', tt_h2))
```

- Table 2.  Summary of variants called by synteny comparison.

```{r echo = FALSE}
kable(tv, format = 'latex', booktabs = T, 
    align = c(rep('l',2),rep('r',ncol(tv)-2)),
    col.names = tv_h0) %>%
  column_spec(1, bold=T) %>%
  #column_spec(2, width_min='2.2cm') %>%
  kable_styling(latex_options = c('striped','scale_down')) %>%
  add_header_above(c(rep('',2),tv_h1)) %>%
  collapse_rows(columns = 1, latex_hline = 'major', valign = 'top')
```

- plot of synteny content and variant density
  - [Mo17 vs B73](/data/11_vnt_density/05.Mo17.pdf)
  - [W22 vs B73](/data/11_vnt_density/05.W22.pdf)
  - [PH207 vs B73](/data/11_vnt_density/05.PH207.pdf)
  - [PHB47 vs B73](/data/11_vnt_density/05.PHB47.pdf)

- Table 3.  Summary of variant effects on syntenic genes.

```{r echo = FALSE}
kable(ef, format = 'latex', booktabs = T,
    align = c(rep('l',2),rep('r',ncol(ef)-3)),
    col.names = ef_h0) %>%
  column_spec(1, bold=T, width_min='1cm') %>%
  kable_styling(latex_options = c('striped','scale_down')) %>%
  add_header_above(c(rep('',2),ef_h1)) %>%
  add_header_above(c(rep('',2),ef_h2)) %>%
  collapse_rows(columns = 1, latex_hline = 'major', valign = 'top')
```
